sklearn/model_selection/_split.py

```diff
@@ -0,0 +1,1388 @@
+
+    See also
+    --------
+    StratifiedKFold
+        For taking label information into account to avoid building folds with
+        imbalanced class distributions (for binary or multiclass
+        classification tasks).
+
+    LabelKFold: K-fold iterator variant with non-overlapping labels.
+    """
+
+    def __init__(self, n_folds=3, shuffle=False,
+                 random_state=None):
+        super(KFold, self).__init__(n_folds, shuffle, random_state)
+        self.shuffle = shuffle
+
+    def _iter_test_indices(self, X, y=None, labels=None):
+        n = _num_samples(X)
+        idxs = np.arange(n)
+        if self.shuffle:
+            check_random_state(self.random_state).shuffle(idxs)
+
+        n_folds = self.n_folds
+        fold_sizes = (n // n_folds) * np.ones(n_folds, dtype=np.int)
+        fold_sizes[:n % n_folds] += 1
+        current = 0
+        for fold_size in fold_sizes:
+            start, stop = current, current + fold_size
+            yield idxs[start:stop]
+            current = stop
+
+
+class LabelKFold(_BaseKFold):
+    """K-fold iterator variant with non-overlapping labels.
+
+    The same label will not appear in two different folds (the number of
+    distinct labels has to be at least equal to the number of folds).
+
+    The folds are approximately balanced in the sense that the number of
+    distinct labels is approximately the same in each fold.
+
+    Parameters
+    ----------
+    n_folds : int, default=3
+        Number of folds. Must be at least 2.
+
+    Examples
+    --------
+    >>> from sklearn.model_selection import LabelKFold
+    >>> X = np.array([[1, 2], [3, 4], [5, 6], [7, 8]])
+    >>> y = np.array([1, 2, 3, 4])
+    >>> labels = np.array([0, 0, 2, 2])
+    >>> label_kfold = LabelKFold(n_folds=2)
+    >>> label_kfold.get_n_splits(X, y, labels)
+    2
+    >>> print(label_kfold)
+    LabelKFold(n_folds=2)
+    >>> for train_index, test_index in label_kfold.split(X, y, labels):
+    ...     print("TRAIN:", train_index, "TEST:", test_index)
+    ...     X_train, X_test = X[train_index], X[test_index]
+    ...     y_train, y_test = y[train_index], y[test_index]
+    ...     print(X_train, X_test, y_train, y_test)
+    ...
+    TRAIN: [0 1] TEST: [2 3]
+    [[1 2]
+     [3 4]] [[5 6]
+     [7 8]] [1 2] [3 4]
+    TRAIN: [2 3] TEST: [0 1]
+    [[5 6]
+     [7 8]] [[1 2]
+     [3 4]] [3 4] [1 2]
+
+    See also
+    --------
+    LeaveOneLabelOut
+        For splitting the data according to explicit domain-specific
+        stratification of the dataset.
+    """
+    def __init__(self, n_folds=3):
+        super(LabelKFold, self).__init__(n_folds, shuffle=False,
+                                         random_state=None)
+
+    def _iter_test_indices(self, X, y, labels):
+        unique_labels, labels = np.unique(labels, return_inverse=True)
+        n_labels = len(unique_labels)
+
+        if self.n_folds > n_labels:
```

Comment:
odd. I got the line as not covered. Could you run coverage after adding the tests I asked for and check?